### PR TITLE
core, server: core version should be stored in the header of the request 

### PR DESCRIPTION
**Pull request by ad-tra**
## Approach 
Changes were made on two fronts:
1. In the routes handler located in `router_service.rs` >  `core_req!`, a warp header filter was added for an `Accept` header which stores the core version. 
2. the `Requester` in `api_service.rs` was changed to inject the core version as an `Accept` header. As far as I understand, most core requests sent to the server pass through  `fn request` in `impl Requester for Network`

### Tests
no tests were written. I'd love to have a discussion about the granularity of tests that we maintain. @Parth suggested a test for requests missing an `Accept` header, but I don't know if or why that's necessary. isn't the purpose of warp filters to validate the request? 

fixes: #1239

### CODE REVIEW

**Review by Parth — CHANGES_REQUESTED** [review]
This is good, but I don't think [accept](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept) makes sense. I didn't look too hard but it looks like there's a [specific schema](https://dotnettutorials.net/lesson/web-api-versioning-using-accept-header/) around `Accept`. I'm happy to just use a header called `core_version` so we don't have to fuss around with this.

I'm going back on my word a bit, but I thought about it a bit and for now, this header needs to be optional. There's a few ways to do this, including [getting all the headers](https://users.rust-lang.org/t/how-to-get-all-the-headers-in-warp/75142/2) and then checking if a particular one exists. Or you can try to: [map the header](https://github.com/seanmonstar/warp/issues/177) to some type and take in that type (prob `Option<String>`).

And finally, work with @SmailBarkouch and get some metrics in here

**Comment by ad-tra** [comment]
> I don't think [accept](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept) makes sense.

@Parth Do you want to use `Accept-Version`, it's not standard but still [commonly used](https://cloud.google.com/blog/products/api-management/api-design-which-version-of-versioning-is-right-for-you)

> There's no standard request header that's ideal for the client to say what format version it wants. The standard "Accept" header specifies which media types the client can accept (e.g., json, yaml, xml, html, plain text), and the standard "Accept-Language" header denotes which natural languages the client can accept (e.g., French, English, Spanish). Some API designers (e.g., the authors of the Restify framework) use a non-standard header called "Accept-Version". If you're doing format versioning, I recommend this header.

If not, we cause use `Core-Version` uppercase, and kebab case is usually the convention for HTTP headers

**Comment by ad-tra** [comment]
What's the line of reasoning behind making the version header optional?

How to handle the missing version header? Do we 
1. polyfill using the version specified in the request body to perform the version check. 
2. or ignore the version check altogether? 

I'm heading towards approach (2)

**Comment by ad-tra** [comment]
@SmailBarkouch is a `HistogramVec` appropriate in this context. Ideally, it would be a pie chart.

**Comment by Parth** [comment]
I'm on board with accept-version.

All I'm looking to do with making the header optional for now, so that people with older versions of core are given a message to upgrade that actually makes sense to them -- otherwise, they'll just get `Unexpected(BadRequest)` or something like that. So what you have now is fine, all you have to change is: in the `None` case, return `ClientUpdateRequired`. After we see, via metrics, that everyone has upgraded, we can elect to make this a required field if desired. 

So all you have to change is:

delete this line:

```rust
        None => Ok(()),
```

(and add some metrics)

**Comment by SmailBarkouch** [comment]
> @SmailBarkouch is a `HistogramVec` appropriate in this context. Ideally, it would be a pie chart.

My knowledge of prometheus's histograms is somewhat limited, but you should be able to achieve a pie chart using a `CounterVec` (what you have right now) or a `GaugeVec`. These series of values can be expressed in Grafana to produce a pie chart pretty effortlessly.

File: server/server/src/lib.rs [hunk 1]
```diff
@@ -57,11 +57,16 @@ macro_rules! internal {
 }
 
 pub fn verify_client_version<Req: Request>(
```
Comment (Parth) [inline server/server/src/lib.rs:59]:
```suggestion
pub fn handle_version<Req: Request>(
```

to suggest that this fn has a side effect

File: server/server/src/router_service.rs [hunk 2]
```diff
@@ -25,6 +27,12 @@ lazy_static! {
         &["request"]
     )
     .unwrap();
+    pub static ref CORE_VERSION_COUNTER: CounterVec = register_counter_vec!(
+        "lockbook_server_core_version",
+        "Lockbook server's supplied core version",
```
Comment (Parth) [inline server/server/src/router_service.rs:32]:
```suggestion
        "Core version request attempts",
```

**Review by Parth — CHANGES_REQUESTED** [review]

**Review by Parth — CHANGES_REQUESTED** [review]
a test that proves that current cores will be able to be asked to update (and will not get a `BadRequest`).

**Review by Parth — APPROVED** [review]

**Pull request merged**